### PR TITLE
pstl][tag_dispatch]  + tag-based dispatching for the algorithms based on __parallel_reduce pattern

### PR DESCRIPTION
**Pull request by MikeDvorskiy**
Tag-based dispatching for the algorithms based on __parallel_reduce pattern: 
adjacent_find, count, count_if, unique, is_partitioned, partition, stable_partition, is_sorted_until, is_sorted, min_element, minmax_element.

### CODE REVIEW

File: pstl/include/pstl/internal/algorithm_impl.h [hunk 1]
```diff
@@ -1752,25 +1758,28 @@ __brick_is_partitioned(_RandomAccessIterator __first, _RandomAccessIterator __la
...
+                         _RandomAccessIterator __last, _UnaryPredicate __pred)
 {
     if (__first == __last)
     {
         return true;
     }
     else
     {
+        using __backend_tag = typename decltype(__tag)::__backend_tag;
+        auto __is_vector = _IsVector{};
```
Comment (rarutyun) [inline pstl/include/pstl/internal/algorithm_impl.h]: You may construct it in-place everywhere you need that. Thus, you don't need a capture.

I would prefer to have to consistent approach. This comment is applicable to other places.

File: pstl/include/pstl/internal/execution_impl.h [hunk 2]
```diff
@@ -95,16 +95,14 @@ using __allow_parallel =
     typename __internal::__policy_traits<typename std::decay<_ExecutionPolicy>::type>::__allow_parallel;
 
 template <typename _ExecutionPolicy, typename... _IteratorTypes>
-typename std::conjunction<__allow_vector<_ExecutionPolicy>,
-                          __is_random_access_iterator<_IteratorTypes>...>::type
+typename std::conjunction<__allow_vector<_ExecutionPolicy>, __is_random_access_iterator<_IteratorTypes>...>::type
```
Comment (rarutyun) [inline pstl/include/pstl/internal/execution_impl.h]: Those changes seem unrelated to this commit. Likely it's clang-format applying.

File: pstl/include/pstl/internal/glue_algorithm_impl.h [hunk 3]
```diff
@@ -185,23 +182,21 @@ __pstl::__internal::__enable_if_execution_policy<_ExecutionPolicy,
...
 template <class _ExecutionPolicy, class _ForwardIterator, class _Predicate>
 __pstl::__internal::__enable_if_execution_policy<_ExecutionPolicy,
                                                  typename iterator_traits<_ForwardIterator>::difference_type>
 count_if(_ExecutionPolicy&& __exec, _ForwardIterator __first, _ForwardIterator __last, _Predicate __pred)
 {
-    return __pstl::__internal::__pattern_count(
-        std::forward<_ExecutionPolicy>(__exec), __first, __last, __pred,
-        __pstl::__internal::__is_parallelization_preferred<_ExecutionPolicy, _ForwardIterator>(__exec),
-        __pstl::__internal::__is_vectorization_preferred<_ExecutionPolicy, _ForwardIterator>(__exec));
+    auto __dispatch_tag = __pstl::__internal::__select_backend(__exec, __first);
```
Comment (rarutyun) [inline pstl/include/pstl/internal/glue_algorithm_impl.h:192]: Nit: Just for the sake of consistency I would prefer to see extra empty line between the line with getting tag and the next one. But it's the least important comment I could do.

File: pstl/include/pstl/internal/algorithm_impl.h [hunk 4]
```diff
@@ -1141,26 +1141,28 @@ __brick_count(_ForwardIterator __first, _ForwardIterator __last, _Predicate __pr
...
     typedef typename std::iterator_traits<_RandomAccessIterator>::difference_type _SizeType;
     return __internal::__except_handler([&]() {
         return __par_backend::__parallel_reduce(
-            std::forward<_ExecutionPolicy>(__exec), __first, __last, _SizeType(0),
-            [__pred, __is_vector](_RandomAccessIterator __begin, _RandomAccessIterator __end, _SizeType __value)
-                -> _SizeType { return __value + __internal::__brick_count(__begin, __end, __pred, __is_vector); },
+            __backend_tag{}, std::forward<_ExecutionPolicy>(__exec), __first, __last, _SizeType(0),
+            [__pred, __is_vector](_RandomAccessIterator __begin, _RandomAccessIterator __end,
+                                  _SizeType __value) -> _SizeType {
+                return __value + __internal::__brick_count(__begin, __end, __pred, typename __backend_tag::_IsVector{});
```
Comment (rarutyun) [inline pstl/include/pstl/internal/algorithm_impl.h]: I believe this is an error.
Should be
```suggestion
                return __value + __internal::__brick_count(__begin, __end, __pred, _IsVector{});
```

File: pstl/include/pstl/internal/algorithm_impl.h [hunk 5]
```diff
@@ -1148,26 +1148,28 @@ __brick_count(_ForwardIterator __first, _ForwardIterator __last, _Predicate __pr
...
 
-template <class _ExecutionPolicy, class _ForwardIterator, class _Predicate, class _IsVector>
+template <class _Tag, class _ExecutionPolicy, class _ForwardIterator, class _Predicate>
 typename std::iterator_traits<_ForwardIterator>::difference_type
-__pattern_count(_ExecutionPolicy&&, _ForwardIterator __first, _ForwardIterator __last, _Predicate __pred,
-                /* is_parallel */ std::false_type, _IsVector __is_vector) noexcept
+__pattern_count(_Tag, _ExecutionPolicy&&, _ForwardIterator __first, _ForwardIterator __last, _Predicate __pred) noexcept
 {
-    return __internal::__brick_count(__first, __last, __pred, __is_vector);
+    return __internal::__brick_count(__first, __last, __pred, typename _Tag::_IsVector{});
```
Comment (rarutyun) [inline pstl/include/pstl/internal/algorithm_impl.h]: Should be:
```suggestion
    return __internal::__brick_count(__first, __last, __pred, typename _Tag::__is_vector{});
```
here and everywhere else. Otherwise, compilation error

File: pstl/include/pstl/internal/algorithm_impl.h (hunk 1 above)
Comment (MikeDvorskiy) [reply pstl/include/pstl/internal/algorithm_impl.h]: Not quite agree. 
Of course, everyone it can be in-placed. But, I did it deliberately to minimize the changes(in case of many occurrences of __is_vector), especially for more comfortable review process by guys from LLVM.
With performance perspective there is no difference, I believe.

File: pstl/include/pstl/internal/execution_impl.h (hunk 2 above)
Comment (MikeDvorskiy) [reply pstl/include/pstl/internal/execution_impl.h]: yes.. I'll redo that.

File: pstl/include/pstl/internal/glue_algorithm_impl.h (hunk 3 above)
Comment (MikeDvorskiy) [reply pstl/include/pstl/internal/glue_algorithm_impl.h:192]: I think it should be done "case by case". Yes, some times an empty line makes readability better, otherwise we have a messy  form code lines.... But, if a function (or scope) contains just two lines, adding adding an empty line between is redundant, I think.

File: pstl/include/pstl/internal/algorithm_impl.h (hunk 4 above)
Comment (MikeDvorskiy) [reply pstl/include/pstl/internal/algorithm_impl.h]: BTW, In my opinion, `__backend_tag::__is_vector{}` is more right approach with architecture perspective.
The type `__backend_tag::__is_vector` is a contract for backend tag type.
Type _IsVector is kind of implementation derails of "__parallel_tag".
Moreover, writing as  `__backend_tag::__is_vector()` instead of `__backend_tag::__is_vector{}` is more preferable, because it will work in case of function __is_vector() as well, if the implementation details will be changed. And function is more general.

File: pstl/include/pstl/internal/algorithm_impl.h (hunk 1 above)
Comment (rarutyun) [reply pstl/include/pstl/internal/algorithm_impl.h]: The goal is not to minimize the changes. They are already huge. I believe we need to avoid capturing when we can. But this is really small thing.

I am more concerned about consistency in terms of one patch to be submitted in LLVM.

File: pstl/include/pstl/internal/glue_algorithm_impl.h (hunk 3 above)
Comment (rarutyun) [reply pstl/include/pstl/internal/glue_algorithm_impl.h:192]: As I said this comment may be completely ignored.

File: pstl/include/pstl/internal/algorithm_impl.h (hunk 4 above)
Comment (rarutyun) [reply pstl/include/pstl/internal/algorithm_impl.h]: Functions with some runtime don't work because we cannot change the return type at run-time. We agreed that all tags of the first level should have `__is_vector` and some of them could have `backend_tag` if necessary

**Review by dparanic — COMMENTED** [review]
LGTM

**Review by rarutyun — APPROVED** [review]
